File: src/main/python/AoC2015_02.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 2
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Present = tuple[int, int, int]
Input = list[Present]
Output1 = int
Output2 = int


TEST1 = "2x3x4"
TEST2 = "1x1x10"


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        def parse_present(line: str) -> Present:
            lg, w, h = map(int, line.split("x"))
            return (lg, w, h)

        return [parse_present(line) for line in input_data]

    def part_1(self, presents: Input) -> Output1:
        def calculate_required_area(present: Present) -> int:
            lg, w, h = present
            sides = (2 * lg * w, 2 * w * h, 2 * h * lg)
            return sum(sides) + min(sides) // 2

        return sum(calculate_required_area(present) for present in presents)

    def part_2(self, presents: Input) -> Output2:
        def calculate_required_length(present: Present) -> int:
            lg, w, h = present
            circumferences = (2 * (lg + w), 2 * (w + h), 2 * (h + lg))
            return min(circumferences) + lg * w * h

        return sum(calculate_required_length(present) for present in presents)

    @aoc_samples(
        (
            ("part_1", TEST1, 58),
            ("part_1", TEST2, 43),
            ("part_2", TEST1, 34),
            ("part_2", TEST2, 14),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2015, 2)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/julia/AoC2015_02.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2015_02

include("aoc.jl")

function solve(input, f)
    ans = 0
    for line ∈ input
        l, w, h = parse.(Int, split(line, "x"))
        ans += f(l, w, h)
    end
    return ans
end

function part1(input)
    function area(l, w, h)
        sides = (2 * l * w, 2 * w * h, 2 * h * l)
        return sum(sides) + minimum(sides) ÷ 2
    end
    return solve(input, (l, w, h) -> area(l, w, h))
end

function part2(input)
    function length(l, w, h)
        circumferences = (2 * (l + w), 2 * (w + h), 2 * (h + l))
        return minimum(circumferences) + l * w * h
    end
    return solve(input, (l, w, h) -> length(l, w, h))
end

TEST1 = @aoc_splitlines("2x3x4")
TEST2 = @aoc_splitlines("1x1x10")

function samples()
    @assert part1(TEST1) == 58
    @assert part1(TEST2) == 43
    @assert part2(TEST1) == 34
    @assert part2(TEST2) == 14
end

end # module AoC2015_02

aoc_main(@__FILE__, ARGS, 2015, 2)
